sklearn/utils/_testing.py

```diff
@@ -87,6 +88,13 @@
 assert_raises_regexp = assert_raises_regex
 
 
+DTYPE_TOLERANCES = {
+    #            rtol, atol
+    np.float32: (1e-6, 1e-5),
+    np.float64: (1e-7, 1e-7),
```

Comment:
I think 1e-6 is way too ambitious. We have many tests that won't pass with that. I think it should be 1e-4. (we are not testing setting 1e-15 for float64 afterall).

I think we should have atol=0 by default. atol only make sense to compare against zero. In addition it represents what value can be considered to be zero which depends on the magnitude of the non zeros entries of the array.

Reply:
> I think we should have atol=0 by default. atol only make sense to compare against zero. In addition it represents what value can be considered to be zero which depends on the magnitude of the non zeros entries of the array.

But then this will make it harder to write atol tests that are dtype aware, no? Do you expect test writers to do something like the following?

```python
# b has some zero entries
assert_allclose(a, b, atol=1e-4 if dtype == np.float32 else 1e-7)
```

Reply:
Since atol should only be used in very specific cases, I think it's acceptable.

Having a non zero default for atol, let's say 1e-7, leads to:
```py
assert_allclose(1e-7, 2e-7)
# no error
```
It's equivalent to using `assert_array_almost_equal` which is what I (we ?) fight against since I joined the project.
non-zero values should be compared using relative difference. Comparing against zero requires absolute difference.

Reply:
Alright, thanks for the clarification. I think once this PR is merged, we should write guidelines on how to properly use `assert_allclose` in our test suite.

At the moment we do not mention it in our doc: https://scikit-learn.org/dev/search.html?q=assert_allclose

Reply:
0d42609 added tests, adapted the relative tolerances, and set the absolute tolerance to 0 by default.